arrow-libs/core/arrow-core/src/commonMain/kotlin/arrow/core/raise/Fold.kt

```diff
@@ -152,6 +153,60 @@ public inline fun <Error, A, B> fold(
   }
 }
 
+/**
+ * Execute the [Raise] context function resulting in [A] or any _logical error_ of type [OtherError],
+ * and transform any raised [OtherError] into [Error], which is raised to the outer [Raise].
+ *
+ * <!--- INCLUDE
+ * import arrow.core.Either
+ * import arrow.core.raise.either
+ * import arrow.core.raise.withError
+ * import io.kotest.matchers.shouldBe
+ * -->
+ * ```kotlin
+ * fun test() {
+ *   either<Int, String> {
+ *     withError(String::length) {
+ *       raise("failed")
+ *     }
+ *   } shouldBe Either.Left(6)
+ * }
+ * ```
+ * <!--- KNIT example-raise-dsl-11.kt -->
+ * <!--- TEST lines.isEmpty() -->
+ */
+@RaiseDSL
+@OptIn(DelicateRaiseApi::class)
+@Suppress("UNCHECKED_CAST")
+public inline fun <Error, OtherError, A> Raise<Error>.withError(
+  transform: (OtherError) -> Error,
+  @BuilderInference block: Raise<OtherError>.() -> A
+): A {
+  contract {
+    callsInPlace(block, EXACTLY_ONCE)
+    callsInPlace(transform, AT_MOST_ONCE)
+  }
+  // recover({ return block(this) }) { raise(transform(it)) }
+  val raise = DefaultRaise(false)
```

Comment:
Doesn't this mean that `innerException` would always have no stack trace?

Reply:
Why would that be the case? As I understand the inner stack trace comes from the point where the `RaiseCancellationException` is first thrown.

Reply:
Well, the inner exception is always going to be a `NoTrace` because it's created by a `DefaultRaise(false)`. Otherwise, we would be creating a stack trace every time, which would defeat our optimizations.

Reply:
That is definitely true!